internal/lookout/db-gen/main.go

```diff
@@ -0,0 +1,231 @@
+package main
+
+import (
+	"database/sql"
+	"flag"
+	"fmt"
+	"math/rand"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	_ "github.com/jackc/pgx/v4/stdlib"
+)
+
+const (
+	QUEUED    int = 1
+	PENDING   int = 2
+	RUNNING   int = 3
+	SUCCEEDED int = 4
+	FAILED    int = 5
+	CANCELLED int = 6
+)
+
+var (
+	ClustersN int = 5
+	QueuesN   int = 10
+	OwnersN   int = 100
+	JobSetsN  int = 100
+	NodesN    int = 100
+)
+
+type JobRun struct {
+	runId            string
+	jobId            string
+	cluster          string
+	node             string
+	podNumber        int
+	unableToSchedule string
+}
+
+type Job struct {
+	jobId      string
+	queue      string
+	owner      string
+	jobset     string
+	priority   int
+	submitted  time.Time
+	state      int
+	jobUpdated time.Time
+}
+
+var wg sync.WaitGroup
+
+func main() {
+	rand.Seed(time.Now().UnixNano())
+
+	JobsN := flag.Int("jobs", 10000, "number of jobs to be generated")
+
+	db := setupDB()
+	defer db.Close()
+
+	for i := 0; i < *JobsN; i++ {
+		job_id := uuid.New().String()[:32]
+		newJob := Job{
+			jobId:     job_id,
+			queue:     concat("queue-", QueuesN),
+			owner:     concat("owner-", OwnersN),
+			jobset:    concat("job-set-", JobSetsN),
+			priority:  rand.Intn(10),
+			submitted: Now(),
+			state:     QUEUED,
+		}
+		newJobRun := JobRun{
+			runId:            uuid.NewString()[:32],
+			jobId:            job_id,
+			cluster:          concat("Cluster-", ClustersN),
+			node:             concat("node-", NodesN),
+			podNumber:        rand.Intn(10),
+			unableToSchedule: "false",
+		}
+
+		insert_job := `INSERT INTO job (job_id, queue, owner, jobset, priority, submitted, job_updated, state) VALUES ($1, $2, $3, $4, $5, $6, $7, $8);`
+		insert_job_run := `INSERT INTO job_run (run_id, job_id, cluster, node, pod_number, unable_to_schedule) VALUES ($1, $2, $3, $4, $5, $6);`
+
+		_, err := db.Exec(insert_job, newJob.jobId, newJob.queue, newJob.owner, newJob.jobset, newJob.priority, newJob.submitted, newJob.jobUpdated, newJob.state)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Couldn't insert into job table\n%v\n", err)
+			continue
+		}
+
+		_, err = db.Exec(insert_job_run, newJobRun.runId, newJobRun.jobId, newJobRun.cluster, newJobRun.node, newJobRun.podNumber, newJobRun.unableToSchedule)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Couldn't insert into job_run table\n%v\n", err)
+			continue
+		}
+		wg.Add(1)
+		go updateJobState(&job_id, db)
+	}
+	wg.Wait()
+}
+
+/* Update job state from Queued -> Pending -> Running -> pick random jobs to either fail or cancel, otherwise succeed */
+func updateJobState(job_id *string, db *sql.DB) {
+	defer wg.Done()
+
+	pending(job_id, db)
+	running(job_id, db)
+	switch r := rand.Intn(6); r {
+	case 0:
+		fail(job_id, db)
+	case 1:
+		cancel(job_id, db)
+	default:
+		success(job_id, db)
+	}
+}
+
+func pending(job_id *string, db *sql.DB) {
```

Comment:
I think you should combine these functions into a single function..

func updateAndSetByState(state int, job_id *string, db *sql.DB) {
...
	updateState := `UPDATE job SET state = $1, job_updated = $2 WHERE job_id = $3;`
	_, err := db.Exec(updateState, state, updateTime, job_id)
}

and then call `updateAndSetByState(PENDING, ..)

Most of these functions only differ by State so I'd be nice to combine these.

Reply:
yes, the query for different differs slightly. I refactored it now.